From patchwork Thu Feb 21 09:48:02 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Stefan Raspl <raspl@linux.ibm.com>
X-Patchwork-Id: 10823415
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 84F21139A
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 21 Feb 2019 09:48:14 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 6F6FC2FEC0
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 21 Feb 2019 09:48:14 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 5FD8B302A5; Thu, 21 Feb 2019 09:48:14 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.9 required=2.0 tests=BAYES_00,MAILING_LIST_MULTI,
	RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id DC60E2FEC0
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 21 Feb 2019 09:48:13 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726903AbfBUJsM (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 21 Feb 2019 04:48:12 -0500
Received: from mx0a-001b2d01.pphosted.com ([148.163.156.1]:33414 "EHLO
        mx0a-001b2d01.pphosted.com" rhost-flags-OK-OK-OK-OK)
        by vger.kernel.org with ESMTP id S1725978AbfBUJsM (ORCPT
        <rfc822;kvm@vger.kernel.org>); Thu, 21 Feb 2019 04:48:12 -0500
Received: from pps.filterd (m0098399.ppops.net [127.0.0.1])
        by mx0a-001b2d01.pphosted.com (8.16.0.27/8.16.0.27) with SMTP id
 x1L9iE9X115713
        for <kvm@vger.kernel.org>; Thu, 21 Feb 2019 04:48:11 -0500
Received: from e06smtp03.uk.ibm.com (e06smtp03.uk.ibm.com [195.75.94.99])
        by mx0a-001b2d01.pphosted.com with ESMTP id 2qssje09ap-1
        (version=TLSv1.2 cipher=AES256-GCM-SHA384 bits=256 verify=NOT)
        for <kvm@vger.kernel.org>; Thu, 21 Feb 2019 04:48:10 -0500
Received: from localhost
        by e06smtp03.uk.ibm.com with IBM ESMTP SMTP Gateway: Authorized Use
 Only! Violators will be prosecuted
        for <kvm@vger.kernel.org> from <raspl@linux.ibm.com>;
        Thu, 21 Feb 2019 09:48:08 -0000
Received: from b06cxnps4076.portsmouth.uk.ibm.com (9.149.109.198)
        by e06smtp03.uk.ibm.com (192.168.101.133) with IBM ESMTP SMTP Gateway:
 Authorized Use Only! Violators will be prosecuted;
        (version=TLSv1/SSLv3 cipher=AES256-GCM-SHA384 bits=256/256)
        Thu, 21 Feb 2019 09:48:06 -0000
Received: from b06wcsmtp001.portsmouth.uk.ibm.com
 (b06wcsmtp001.portsmouth.uk.ibm.com [9.149.105.160])
        by b06cxnps4076.portsmouth.uk.ibm.com (8.14.9/8.14.9/NCO v10.0) with
 ESMTP id x1L9m5kM27984006
        (version=TLSv1/SSLv3 cipher=DHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Thu, 21 Feb 2019 09:48:05 GMT
Received: from b06wcsmtp001.portsmouth.uk.ibm.com (unknown [127.0.0.1])
        by IMSVA (Postfix) with ESMTP id C7615A4062;
        Thu, 21 Feb 2019 09:48:05 +0000 (GMT)
Received: from b06wcsmtp001.portsmouth.uk.ibm.com (unknown [127.0.0.1])
        by IMSVA (Postfix) with ESMTP id B5D24A405C;
        Thu, 21 Feb 2019 09:48:05 +0000 (GMT)
Received: from tuxmaker.boeblingen.de.ibm.com (unknown [9.152.85.9])
        by b06wcsmtp001.portsmouth.uk.ibm.com (Postfix) with ESMTPS;
        Thu, 21 Feb 2019 09:48:05 +0000 (GMT)
Received: from tuxmaker.boeblingen.de.ibm.com (localhost [127.0.0.1])
        by tuxmaker.boeblingen.de.ibm.com (Postfix) with ESMTP id 51FA220F5F6;
        Thu, 21 Feb 2019 10:48:05 +0100 (CET)
From: Stefan Raspl <raspl@linux.ibm.com>
To: kvm@vger.kernel.org
Cc: drjones@redhat.com, pbonzini@redhat.com, sergeyb@bronevichok.ru
Subject: [PATCH v1 1/2] tap13: list testcases individually
Date: Thu, 21 Feb 2019 10:48:02 +0100
X-Mailer: git-send-email 2.16.4
In-Reply-To: <20190221094803.56214-1-raspl@linux.ibm.com>
References: <20190221094803.56214-1-raspl@linux.ibm.com>
X-TM-AS-GCONF: 00
x-cbid: 19022109-0012-0000-0000-000002F86371
X-IBM-AV-DETECTION: SAVI=unused REMOTE=unused XFE=unused
x-cbparentid: 19022109-0013-0000-0000-0000212FF862
Message-Id: <20190221094803.56214-2-raspl@linux.ibm.com>
X-Proofpoint-Virus-Version: vendor=fsecure engine=2.50.10434:,,
 definitions=2019-02-21_06:,,
 signatures=0
X-Proofpoint-Spam-Details: rule=outbound_notspam policy=outbound score=0
 priorityscore=1501
 malwarescore=0 suspectscore=1 phishscore=0 bulkscore=0 spamscore=0
 clxscore=1011 lowpriorityscore=0 mlxscore=0 impostorscore=0
 mlxlogscore=999 adultscore=0 classifier=spam adjust=0 reason=mlx
 scancount=1 engine=8.0.1-1810050000 definitions=main-1902210072
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

The current tap13 support summarizes testcases on a rather coarse
granularity. Which sort of defeats the purpose, especially in CI
environments.
This patch improves things by post-processing the log files, extracting
results for individual testcases. One could argue that post-processing the
logs is a rather fragile approach. Which is true - but apparently already
the case, see e.g. extract_summary in scripts/runtime.bash.
Plus this is quite cheap, while a proper solution would require to modify
the kernels, so we can e.g. pass in a counter for the testcase number. Plus
we would probably have to come up with reasonable testcase names, while
this approach simply derives them from each test output.

Signed-off-by: Stefan Raspl <raspl@linux.ibm.com>
---
 run_tests.sh         | 28 +++++++++++++++++++++++-----
 scripts/runtime.bash | 10 ----------
 2 files changed, 23 insertions(+), 15 deletions(-)

diff --git a/run_tests.sh b/run_tests.sh
index 102c806..7ae37e0 100755
--- a/run_tests.sh
+++ b/run_tests.sh
@@ -107,10 +107,6 @@ mkdir $unittest_log_dir || exit 2
 
 echo "BUILD_HEAD=$(cat build-head)" > $unittest_log_dir/SUMMARY
 
-if [[ $tap_output == "yes" ]]; then
-    test_number=0
-    echo "TAP version 13"
-fi
 trap "wait; exit 130" SIGINT
 for_each_unittest $config run_task
 
@@ -118,5 +114,27 @@ for_each_unittest $config run_task
 wait
 
 if [[ $tap_output == "yes" ]]; then
-    echo "1..$test_number"
+    echo "TAP version 13"
+    echo "1..`cat $unittest_log_dir/*.log | grep -E "^PASS|^FAIL|^SKIP" | wc -l`";
+    i=1
+    for log in `ls -1 $unittest_log_dir/*.log`; do
+        while read -r line; do
+            name="`echo ${line% ==*} | cut -c 7-`"
+            case "${line:0:4}" in
+                PASS)
+                    echo "ok $i - $name"
+                    ;;
+                FAIL)
+                    echo "not ok $i - $name"
+                    ;;
+                SKIP)
+                    echo "ok $i - $name # skip"
+                    ;;
+                *)
+                    continue
+                    ;;
+            esac
+            (( i++ ))
+        done < "$log"
+    done
 fi
diff --git a/scripts/runtime.bash b/scripts/runtime.bash
index fffa7dd..e7844f0 100644
--- a/scripts/runtime.bash
+++ b/scripts/runtime.bash
@@ -69,16 +69,6 @@ function print_result()
         fi
         return
     fi
-
-    if [[ $status == "FAIL" ]]; then
-        echo "not ok $testname $reason"
-    elif [[ $status == "PASS" ]]; then
-        echo "ok $testname"
-    elif [[ $status == "SKIP" ]]; then
-        echo "ok $testname # SKIP $reason"
-    else
-        echo "not ok # TODO unknown test status"
-    fi
 }
 
 function run()

From patchwork Thu Feb 21 09:48:03 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Stefan Raspl <raspl@linux.ibm.com>
X-Patchwork-Id: 10823417
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id DAF651390
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 21 Feb 2019 09:48:21 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id C906C2FEC0
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 21 Feb 2019 09:48:21 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id BCCC9302A5; Thu, 21 Feb 2019 09:48:21 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.9 required=2.0 tests=BAYES_00,MAILING_LIST_MULTI,
	RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 5CDA32FEC0
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 21 Feb 2019 09:48:21 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727210AbfBUJsU (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 21 Feb 2019 04:48:20 -0500
Received: from mx0a-001b2d01.pphosted.com ([148.163.156.1]:59698 "EHLO
        mx0a-001b2d01.pphosted.com" rhost-flags-OK-OK-OK-OK)
        by vger.kernel.org with ESMTP id S1726184AbfBUJsN (ORCPT
        <rfc822;kvm@vger.kernel.org>); Thu, 21 Feb 2019 04:48:13 -0500
Received: from pps.filterd (m0098393.ppops.net [127.0.0.1])
        by mx0a-001b2d01.pphosted.com (8.16.0.27/8.16.0.27) with SMTP id
 x1L9iHBB123360
        for <kvm@vger.kernel.org>; Thu, 21 Feb 2019 04:48:12 -0500
Received: from e06smtp04.uk.ibm.com (e06smtp04.uk.ibm.com [195.75.94.100])
        by mx0a-001b2d01.pphosted.com with ESMTP id 2qssj80ay6-1
        (version=TLSv1.2 cipher=AES256-GCM-SHA384 bits=256 verify=NOT)
        for <kvm@vger.kernel.org>; Thu, 21 Feb 2019 04:48:11 -0500
Received: from localhost
        by e06smtp04.uk.ibm.com with IBM ESMTP SMTP Gateway: Authorized Use
 Only! Violators will be prosecuted
        for <kvm@vger.kernel.org> from <raspl@linux.ibm.com>;
        Thu, 21 Feb 2019 09:48:09 -0000
Received: from b06cxnps4075.portsmouth.uk.ibm.com (9.149.109.197)
        by e06smtp04.uk.ibm.com (192.168.101.134) with IBM ESMTP SMTP Gateway:
 Authorized Use Only! Violators will be prosecuted;
        (version=TLSv1/SSLv3 cipher=AES256-GCM-SHA384 bits=256/256)
        Thu, 21 Feb 2019 09:48:06 -0000
Received: from d06av22.portsmouth.uk.ibm.com (d06av22.portsmouth.uk.ibm.com
 [9.149.105.58])
        by b06cxnps4075.portsmouth.uk.ibm.com (8.14.9/8.14.9/NCO v10.0) with
 ESMTP id x1L9m5w425952460
        (version=TLSv1/SSLv3 cipher=DHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Thu, 21 Feb 2019 09:48:06 GMT
Received: from d06av22.portsmouth.uk.ibm.com (unknown [127.0.0.1])
        by IMSVA (Postfix) with ESMTP id D655E4C04A;
        Thu, 21 Feb 2019 09:48:05 +0000 (GMT)
Received: from d06av22.portsmouth.uk.ibm.com (unknown [127.0.0.1])
        by IMSVA (Postfix) with ESMTP id C265F4C040;
        Thu, 21 Feb 2019 09:48:05 +0000 (GMT)
Received: from tuxmaker.boeblingen.de.ibm.com (unknown [9.152.85.9])
        by d06av22.portsmouth.uk.ibm.com (Postfix) with ESMTPS;
        Thu, 21 Feb 2019 09:48:05 +0000 (GMT)
Received: from tuxmaker.boeblingen.de.ibm.com (localhost [127.0.0.1])
        by tuxmaker.boeblingen.de.ibm.com (Postfix) with ESMTP id 5C69E20F5F7;
        Thu, 21 Feb 2019 10:48:05 +0100 (CET)
From: Stefan Raspl <raspl@linux.ibm.com>
To: kvm@vger.kernel.org
Cc: drjones@redhat.com, pbonzini@redhat.com, sergeyb@bronevichok.ru
Subject: [kvm-unit-tests PATCH v1 2/2] run_tests: fix command line options
 handling
Date: Thu, 21 Feb 2019 10:48:03 +0100
X-Mailer: git-send-email 2.16.4
In-Reply-To: <20190221094803.56214-1-raspl@linux.ibm.com>
References: <20190221094803.56214-1-raspl@linux.ibm.com>
X-TM-AS-GCONF: 00
x-cbid: 19022109-0016-0000-0000-000002591A77
X-IBM-AV-DETECTION: SAVI=unused REMOTE=unused XFE=unused
x-cbparentid: 19022109-0017-0000-0000-000032B36939
Message-Id: <20190221094803.56214-3-raspl@linux.ibm.com>
X-Proofpoint-Virus-Version: vendor=fsecure engine=2.50.10434:,,
 definitions=2019-02-21_06:,,
 signatures=0
X-Proofpoint-Spam-Details: rule=outbound_notspam policy=outbound score=0
 priorityscore=1501
 malwarescore=0 suspectscore=1 phishscore=0 bulkscore=0 spamscore=0
 clxscore=1015 lowpriorityscore=0 mlxscore=0 impostorscore=0
 mlxlogscore=999 adultscore=0 classifier=spam adjust=0 reason=mlx
 scancount=1 engine=8.0.1-1810050000 definitions=main-1902210072
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

When specifying desired tests on the command line, options are ignored
unless specified prior to any desired tests.
E.g.

  $ ./run_tests.sh sthyi -t emulator -j5
  PASS emulator (102 tests)
  PASS sthyi (19 tests)

Note that the output is not in TAP13 format, and the tests are not executed
in parallel.
This patch fixes the issue by switching to getopt instead. Might not be the
optimum solution, but threw in long command line opts as a bonus.

Signed-off-by: Stefan Raspl <raspl@linux.ibm.com>
---
 run_tests.sh | 45 +++++++++++++++++++++++++++------------------
 1 file changed, 27 insertions(+), 18 deletions(-)

diff --git a/run_tests.sh b/run_tests.sh
index 7ae37e0..a0555ce 100755
--- a/run_tests.sh
+++ b/run_tests.sh
@@ -17,13 +17,13 @@ cat <<EOF
 
 Usage: $0 [-h] [-v] [-a] [-g group] [-j NUM-TASKS] [-t]
 
-    -h: Output this help text
-    -v: Enables verbose mode
-    -a: Run all tests, including those flagged as 'nodefault'
-        and those guarded by errata.
-    -g: Only execute tests in the given group
-    -j: Execute tests in parallel
-    -t: Output test results in TAP format
+    -h, --help      Output this help text
+    -v, --verbose   Enables verbose mode
+    -a, --all       Run all tests, including those flagged as 'nodefault'
+                    and those guarded by errata.
+    -g, --group     Only execute tests in the given group
+    -j, --parallel  Execute tests in parallel
+    -t, --tap13     Output test results in TAP format
 
 Set the environment variable QEMU=/path/to/qemu-system-ARCH to
 specify the appropriate qemu binary for ARCH-run.
@@ -34,36 +34,45 @@ EOF
 RUNTIME_arch_run="./$TEST_DIR/run"
 source scripts/runtime.bash
 
-while getopts "ag:htj:v" opt; do
-    case $opt in
-        a)
+only_tests=""
+args=`getopt -u -o ag:htj:v -l all,group:,help,tap13,parallel:,verbose -- $*`
+[ $? -ne 0 ] && exit 2;
+set -- $args;
+while [ $# -gt 0 ]; do
+    case "$1" in
+        -a | --all)
             run_all_tests="yes"
             export ERRATA_FORCE=y
             ;;
-        g)
-            only_group=$OPTARG
+        -g | --group)
+            shift
+            only_group=$1
             ;;
-        h)
+        -h | --help)
             usage
             exit
             ;;
-        j)
-            unittest_run_queues=$OPTARG
+        -j | --parallel)
+            shift
+            unittest_run_queues=$1
             if (( $unittest_run_queues <= 0 )); then
                 echo "Invalid -j option: $unittest_run_queues"
                 exit 2
             fi
             ;;
-        v)
+        -v | --verbose)
             verbose="yes"
             ;;
-        t)
+        -t | --tap13)
             tap_output="yes"
             ;;
+        --)
+            ;;
         *)
-            exit 2
+            only_tests="$only_tests $1"
             ;;
     esac
+    shift
 done
 
 # RUNTIME_log_file will be configured later
